protocols/autonatv2/src/client/handler.rs

```diff
@@ -0,0 +1,395 @@
+};
+use void::Void;
+
+use crate::{
+    client::ToBehaviour,
+    generated::structs::mod_DialResponse::ResponseStatus,
+    request_response::{
+        DialBack, DialDataRequest, DialDataResponse, DialRequest, DialResponse, Request, Response,
+    },
+    DIAL_BACK_PROTOCOL_NAME, REQUEST_PROTOCOL_NAME, REQUEST_UPGRADE,
+};
+use crate::{DATA_FIELD_LEN_UPPER_BOUND, DATA_LEN_LOWER_BOUND, DATA_LEN_UPPER_BOUND};
+
+use super::ResponseInfo;
+
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(10);
+const MAX_CONCURRENT_REQUESTS: usize = 10;
+
+#[derive(Debug, thiserror::Error)]
+pub(super) enum Error {
+    #[error("io error")]
+    Io(#[from] io::Error),
+    #[error("invalid data request index: {index} (max: {max})")]
+    InvalidDataRequestIndex { index: usize, max: usize },
+    #[error("data request too large: {len} (max: {max})")]
+    DataRequestTooLarge { len: usize, max: usize },
+    #[error("data request too small: {len} (min: {min})")]
+    DataRequestTooSmall { len: usize, min: usize },
+    #[error("timeout")]
+    Timeout(#[from] futures_bounded::Timeout),
+    #[error("invalid nonce: {nonce} (provided by (by))")]
+    WrongNonceGiven { nonce: u64, by: PeerId },
+    #[error("request protocol removed by peer")]
+    RequestProtocolRemoved,
+    #[error("dial back protocol removed by me")]
+    DialBackProtocolRemoved,
+}
+
+pub(super) struct Handler {
+    /// Queue of events to return when polled.
+    queued_events: VecDeque<
+        ConnectionHandlerEvent<
+            <Self as ConnectionHandler>::OutboundProtocol,
+            <Self as ConnectionHandler>::OutboundOpenInfo,
+            <Self as ConnectionHandler>::ToBehaviour,
+            <Self as ConnectionHandler>::Error,
+        >,
+    >,
+    pending_data: VecDeque<PendingData>,
+    queued_requests: VecDeque<DialRequest>,
+    inbound: futures_bounded::FuturesSet<Result<(), Error>>,
+    outbound: futures_bounded::FuturesSet<Result<PerformRequestStatus, Error>>,
+    accepted_nonce: Arc<scc::HashSet<u64>>,
+    remote_peer_id: PeerId,
+}
+
+impl Handler {
+    pub(super) fn new(accepted_nonce: Arc<scc::HashSet<u64>>, remote_peer_id: PeerId) -> Self {
+        Self {
+            queued_events: VecDeque::new(),
+            pending_data: VecDeque::new(),
+            queued_requests: VecDeque::new(),
+            inbound: futures_bounded::FuturesSet::new(DEFAULT_TIMEOUT, MAX_CONCURRENT_REQUESTS),
+            outbound: futures_bounded::FuturesSet::new(DEFAULT_TIMEOUT, MAX_CONCURRENT_REQUESTS),
+            accepted_nonce,
+            remote_peer_id,
+        }
+    }
+}
+
+impl ConnectionHandler for Handler {
+    type Error = Error;
+    type ToBehaviour = super::ToBehaviour<Self::Error>;
+    type FromBehaviour = super::FromBehaviour;
+    type InboundProtocol = ReadyUpgrade<StreamProtocol>;
+    type OutboundProtocol = ReadyUpgrade<StreamProtocol>;
+    type InboundOpenInfo = ();
+    type OutboundOpenInfo = Option<PendingData>;
```

Comment:
`OpenInfo` will be deprecated at some point, we don't want to use it anymore within the workspace.